### ISSUE TITLE
6. Fine-tune pre-trained model for 37-class breed classification

### ISSUE THREAD

**Issue by lizamaa123** [lizamaa123/DD2424_PROJECT_GROUP_11#9]
**_Description:_**
Fine-tune the pre-trained network and compare its performance against linear probing. (Strategy 1 and 2)

**Tasks:**

- [x] unfreeze last layer block first
- [x] train and evaluate
- [x] optionally test unfreezing more layers
- [x] compare validation/test performance to linear probing
- [x] track training time and accuracy

**Definition of done:**

- [ ] at least one fine-tuned model is trained
- [x] comparison to linear probing is documented

**Comment by lizamaa123** [comment]
### STRATEGY 1:
Initially ran into a "forgetting" issue where the model just flatlined at ~2.7% accuracy (random guessing). It turned out the initial learning rate (2e-3) we used for the linear probe was way too big for the pre-trained convolutional weights. Dropping it to lr=1e-4 fixed it perfectly!

Here are the final results for the simultaneous fine-tuning:

**L=1**
- Best test accuracy: **88.1439%**
- Total runtime: 239.25 seconds

**L=2**
- Best test accuracy: **86.2905%**
- Total runtime: 494.52 seconds

**L=3**
- Best test accuracy: **86.3723%**
- Total runtime: 765.41 seconds

**L=4**
- Best test accuracy: **85.5819%**
- Total runtime: 1078.22 seconds

The results indicate that unfreezing just the last block (L=1) gave the best test performance and ran the fastest. As we unfroze deeper into the network (using less and less layers), the model simply overfit the training data faster, resulting in a drop in test accuracy.

Following are the plots in order from L=1 to L=4:

<img width="1200" height="500" alt="Image" src="https://github.com/user-attachments/assets/d13d9e14-bc8e-48dc-a76f-0f825f1ab6d1" />

<img width="1200" height="500" alt="Image" src="https://github.com/user-attachments/assets/5fe1fa89-3445-425f-869b-8c277bb33833" />

<img width="1200" height="500" alt="Image" src="https://github.com/user-attachments/assets/043fe019-6e44-4dd9-b6b5-e558148cb286" />

<img width="1200" height="500" alt="Image" src="https://github.com/user-attachments/assets/3a09a502-a77d-47e6-8a54-ea03c6a466ff" />

**Comment by lizamaa123** [comment]
### STRATEGY 2:
Instead of reinitiating a new random model for every iteration, like in strategy 1, the model stays the but unfreezes following last layers
**L=0 (linear probing)**

- Best test accuracy: **88.2257%**
- Total runtime: 223.68 seconds

**L=1**

- Best test accuracy: **90.2426%**
- Total runtime: 465.70 seconds

**L=2**

- Best test accuracy: **88.0894%**
- Total runtime: 707.72 seconds

**L=3**

- Best test accuracy: **86.5631%**
- Total runtime: 963.67 seconds

**L=4** 

- Best test accuracy: **87.1900%**
- Total runtime: 1234.97 seconds

It was possible to achieve **>90%**, however, there is still a great overfitting as you can see from following plots starting from L=0 to L=4:

<img width="1200" height="500" alt="Image" src="https://github.com/user-attachments/assets/1182eba9-9c15-47be-9a8e-80dd989797dd" />

<img width="1200" height="500" alt="Image" src="https://github.com/user-attachments/assets/844b4cea-bb17-4166-a1f1-90c4775ecd29" />

<img width="1200" height="500" alt="Image" src="https://github.com/user-attachments/assets/c9085f5c-5eb6-470f-839e-bf1e56da7a13" />

<img width="1200" height="500" alt="Image" src="https://github.com/user-attachments/assets/69a121c6-7ae6-42ad-9231-52ce86d88a4e" />

<img width="1200" height="500" alt="Image" src="https://github.com/user-attachments/assets/cf091d9e-7a2e-4302-86c6-2906d6b9ff5d" />
